### PR TITLE
Add Radio (and label) components

### PR DESCRIPTION
**Pull request by nickcolley**
Forked from GOV.UK Frontend

Review app component guide:
https://government-frontend-pr-552.herokuapp.com/component-guide/radio
https://government-frontend-pr-552.herokuapp.com/component-guide/label

Depends on https://github.com/alphagov/govuk_publishing_components/pull/101

Biggest change from GOV.UK Frontend is that this PR makes use of `aria-describedby` for the hint text.

https://www.powermapper.com/tests/screen-readers/labelling/input-text-aria-describedby/

https://trello.com/c/yo9dWRM8/242-create-a-radio-button-component

TODO:

- [ ]  Fix [govuk_publishing_components output of symbols](https://github.com/alphagov/govuk_publishing_components/issues/102)
- [x] Test on real mobile Android + iOS
- [x] Test in [assistive tech](https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices#testing-with-assistive-technologies)
- [x] Test [cross browser](https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices#browsers-to-test-in)
- [x] Add localisation for 'or divider'
- [x] Add component integration tests
- [x] Check if these work when changing colours
- [x] Add interaction integration tests?

## Cross browser testing
Tested that the radios render correctly and are functional.

In the older internet explorers they fall back correctly, and have no barriers.

| Operating system | Browser                             | Tested                                                                                                         |
|------------------|-------------------------------------|----------------------------------------------------------------------------------------------------------------|
| Windows          | Internet Explorer 8-10              | ✅ Windows 7, IE 8 - 10                                                                                           |
|                  | Internet Explorer 11                | ✅ Windows 7, IE 11                                                                                               |
|                  | Edge (latest versions)              | ✅ Windows 10, Edge 16                                                                                            |
|                  | Google Chrome (latest versions)     | ✅ Windows 10, Chrome 62                                                                                          |
|                  | Mozilla Firefox (latest versions)   | ✅ Windows 10, Firefox 57                                                                                         |
| macOS            | Safari 9+                           | ✅ macOS High Sierra, Safari 11- macOS El Capitan, Safari 9.1                                                     |
|                  | Google Chrome (latest versions)     | ✅ macOS High Sierra, Chrome 61                                                                                   |
|                  | Mozilla Firefox (latest versions)   | ✅ macOS High Sierra, Firefox 57                                                                                  |
| iOS              | Safari 9.3+                         | ✅ iPhone 6S, iOS 9, Safari - iPhone 6, iOS 8, Safari - iPhone 7, iOS 10, Safari                                  |
|                  | Google Chrome (latest versions)     | ✅ iPhone 6s, iOS 8, Chrome                                                                                       |
| Android          | Mozilla Firefox (latest versions)   | ✅ Samsung Galaxy S8, Firefox                                                                                     |
|                  | Google Chrome (latest versions)     | ✅ Samsung Galaxy S5, Android 4.4, Chrome - Google Nexus 6, Android 5, Chrome - Google Nexus 7, Android 6, Chrome |
| Windows Phone    | Internet Explorer (latest versions) | ✅ Windows Phone, Lumia 930 8.1                                                                                   |

## Assistive technology testing

Since this version of the radio buttons includes aria-describedby to associate the hint text with the labels, we should make sure to check that this behaves as expected when testing.

| Software                 | Version        | Type               | Browser                                                           | Tested |
|--------------------------|----------------|--------------------|-------------------------------------------------------------------|--------|
| JAWS                     | 15 or later    | screen reader      | Internet Explorer 11                                              | ✅ announces label followed by announcing 'radio button' then announces the hint text followed    |
| ZoomText                 | 10.11 or later | screen magnifier   | Internet Explorer 11                                              | ✅ elements can be seen and understood together    |
| Dragon NaturallySpeaking | 11 or later    | speech recognition | Internet Explorer 11                                              | ✅ works with commands 'tab' to inputs, and also 'click gateway, click government gateway, click gov.uk verify, click verify'   |
| NVDA                     | 2016 or later  | screen reader      | Mozilla Firefox (latest versions)                                 | ✅ announces label followed by announcing 'radio button' then announces the hint text followed   |
| VoiceOver                | n/a            | screen reader      | Safari 10 (or higher) on Mac OS X 10.10, and iOS 10.3 (or higher) | ✅ announces label followed by announcing 'radio button' then announces the hint text followed by a final announcement of 'radio button'    |

### CODE REVIEW

**Comment by nickcolley** [comment]
Good illustration that using the bold feature + hint text could help people who change colours.

screen shot 2017-11-28 at 15 05 06

**Comment by nickcolley** [comment]
I've decided to move the output of the symbols in documentation to an issue and into our backlog so we don't spend too much time.

I'll update this fixture with a description explaining.

https://github.com/alphagov/govuk_publishing_components/issues/102

**Comment by nickcolley** [comment]
This will need a squash once we've got https://github.com/alphagov/govuk_publishing_components/pull/103 in

File: app/views/components/docs/label.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ description: Use labels for all form fields.
 body: |
   For use with other form inputs e.g. [Radio buttons](/component-guide/radio)
 
-  Forked from the upcoming GOV.UK Frontend, when GOV.UK Frontend release we can replace these source files.
+  Forked from the upcoming [GOV.UK Frontend](https://github.com/alphagov/govuk-frontend), when GOV.UK Frontend release we can replace these source files.
```
Comment (leenagupte) [inline app/views/components/docs/label.yml]: I would squash this line edit into commit 543cc481fac6339fcaf22b11460df5247c132bf8 where the line was added.

File: app/views/components/docs/radio.yml [hunk 2]
```diff
@@ -1,11 +1,11 @@
 name: Radio button
 description: A radio button is a GOV.UK element that allows users to answer a question by selecting an option. If you have a question with more than one option you should stack radio buttons.
 body: |
-  Forked from the upcoming GOV.UK Frontend, when GOV.UK Frontend release we can replace these source files.
+  Forked from the upcoming [GOV.UK Frontend](https://github.com/alphagov/govuk-frontend), when GOV.UK Frontend release we can replace these source files.
```
Comment (leenagupte) [inline app/views/components/docs/radio.yml]: I would squash this line edit into commit 543cc481fac6339fcaf22b11460df5247c132bf8 where the line was added.

File: app/views/components/docs/radio.yml [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
 
   This should always be used nested in a fieldset so that it has proper context, useful for users of assistive technologies.
 
-  You can also pass 'or' as an item to [break up radios](https://www.gov.uk/service-manual/design/writing-for-user-interfaces#ask-questions-that-users-can-understand)
+  You can also use 'or' as an item to [break up radios](https://www.gov.uk/service-manual/design/writing-for-user-interfaces#ask-questions-that-users-can-understand)
```
Comment (leenagupte) [inline app/views/components/docs/radio.yml]: I would squash this line update into commit 53425fce1e3af3b971b90206d8280390d6151233 where it was added.

File: app/views/components/docs/radio.yml [hunk 4]
```diff
@@ -0,0 +1,117 @@
+name: Radio button
+description: A radio button is a GOV.UK element that allows users to answer a question by selecting an option. If you have a question with more than one option you should stack radio buttons.
+body: |
+  Forked from the upcoming [GOV.UK Frontend](https://github.com/alphagov/govuk-frontend), when GOV.UK Frontend release we can replace these source files.
+
+  This should always be used nested in a fieldset so that it has proper context, useful for users of assistive technologies.
```
Comment (vanitabarrett) [inline app/views/components/docs/radio.yml]: Is it worth re-iterating this in the accessibility criteria too?

File: app/views/components/_label.html.erb [hunk 5]
```diff
@@ -0,0 +1,40 @@
+<%
+  items ||= []
```
Comment (vanitabarrett) [inline app/views/components/_label.html.erb]: Where is this used? 🤔

File: app/views/components/_label.html.erb [hunk 6]
```diff
@@ -3,14 +3,12 @@
   classes ||= ''
   text_classes ||= ''
   html_for ||= ''
-  text ||= ''
```
[the remark this answers is not in the document]
Comment (nickcolley) [reply app/views/components/_label.html.erb]: The component should fail if no text has been passed to it, this is covered in the tests.

File: app/views/components/docs/radio.yml [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  default:
+    data:
+      name: "radio-group"
+      items:
+      - value: "government-gateway"
+        text: "Use Government Gateway"
+      - value: "govuk-verify"
+        text: "Use GOV.UK Verify"
+  with_bold:
+    description: "Used to provide better contrast between long labels and hint text"
```
Comment (vanitabarrett) [inline app/views/components/docs/radio.yml]: Should this example then include hint text to demonstrate this?

File: app/views/components/docs/label.yml [hunk 8]
```diff
@@ -0,0 +1,25 @@
+name: Form label
+description: Use labels for all form fields.
+body: |
+  For use with other form inputs e.g. [Radio buttons](/component-guide/radio)
+
+  Forked from the upcoming [GOV.UK Frontend](https://github.com/alphagov/govuk-frontend), when GOV.UK Frontend release we can replace these source files.
+accessibility_criteria: |
+  - have a text contrast ratio higher than 4.5:1 against the background colour to meet [WCAG AA](https://www.w3.org/TR/WCAG20/#visual-audio-contrast-contrast)
+  - have visible text
+  - `hint_id` is matched with an `aria-describedby` property on the input your label is assosiated with.
```
Comment (vanitabarrett) [inline app/views/components/docs/label.yml]: "associated"

File: test/components/radio_test.rb [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+class RadioTest < ComponentTestCase
+  def component_name
+    "radio"
+  end
+
+  test "does not render radio when no data is given" do
+    assert_empty render_component({})
+  end
+
+  test "does not render radio when no items but no name is given" do
```
Comment (vanitabarrett) [inline test/components/radio_test.rb]: Not sure this test name is clear... Also can we tweak the name of this test and the previous one to indicate that the first is meant to render nothing and this one is meant to throw an error?

File: app/views/components/docs/radio.yml (hunk 7 above)
Comment (nickcolley) [reply app/views/components/docs/radio.yml]: I was trying to only use the individual features, but it seems a good idea to show both here good point.

**Comment by nickcolley** [comment]
@maxgds there's not really a great way to do that inside the iterated items, I'm thinking we'll cover that possibility in the wider integration tests when we use this component, do you think that'd be ok?

**Comment by maxgds** [comment]
"I'm thinking we'll cover that possibility in the wider integration tests when we use this component"

I'm not sure it is enough because I think further down the line someone may not do the necessary tests when using the component. Also, if the input is being autogenerated from a key/value list somewhere else then that could break and silently give a bad experience to the user. Although silently removing an option or not rendering the list of radios at all is probably not much better now I think about it.

**Comment by nickcolley** [comment]
@maxgds if we were using radio inputs directly, html would also behave the same way that this component does. I agree with you, but feel like putting validation in this component is odd to do and is the responsibility of the rest of the application.

We'd have tests for the page that consume this component to ensure the user journey works, as well as presenter tests to ensure that components get the right data.

**Comment by nickcolley** [comment]
@leenagupte going to try and squash it down, wish me luck. 💀

File: app/views/components/_label.html.erb (hunk 6 above)
[the remark this answers is not in the document]
Comment (leenagupte) [reply app/views/components/_label.html.erb]: Which test covers this? Is it in another commit?

**Comment by nickcolley** [comment]
@leenagupte I've split the commits based on new features added from GOV.UK Frontend, so the base commit is as vanilla a port of GOV.UK Frontend as possible, with the additional commits so they work better for us.

File: test/components/radio_test.rb [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+      render_component(items: [
+        {
+          value: "government-gateway",
+          text: "Use Government Gateway"
+        }
+      ])
+    end
+  end
+
+  test "renders radio-group with one item" do
```
Comment (leenagupte) [inline test/components/radio_test.rb]: Pedantry alert! Should you be allowed to render a radio group with only one item? It rather takes away the illusion of choice.

File: test/components/radio_test.rb [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+    assert_select ".app-c-radio__input[name=radio-group-custom-id-prefix]"
+    assert_select ".app-c-radio__input[id=custom-0]", value: "government-gateway"
+    assert_select ".app-c-radio__label[for=custom-0]", text: "Use Government Gateway"
+    assert_select ".app-c-radio__input[id=custom-1]", value: "govuk-verify"
+    assert_select ".app-c-radio__label[for=custom-1]", text: "Use GOV.UK Verify"
+  end
+
+  # This component can be interacted with, so use integration tests for these cases.
+  class RadioIntegrationTest < ActionDispatch::IntegrationTest
```
Comment (leenagupte) [inline test/components/radio_test.rb]: I think that these tests should be in a separate file under integration tests.

File: test/components/radio_test.rb (hunk 10 above)
Comment (nickcolley) [reply test/components/radio_test.rb]: I suppose not, that test is not really that useful is it. :)

**Comment by leenagupte** [comment]
@nickcolley The rebased commits look really good. 

I think the last commit (2baee91c40076cf3e0d09a266c0e9bbc9bd0d690) needs a bit more explanation. What is the benefit of splitting the hint text from the label? How is it related to renaming/adding(?) the radio-button classes?

File: test/components/radio_test.rb (hunk 11 above)
Comment (nickcolley) [reply test/components/radio_test.rb]: I'd prefer to have tests colocated so they're easy to update and maintain, components don't really fit well with Rails conventions in this respect.

Ideally all component related files would be together, but that's for another time.

**Comment by nickcolley** [comment]
@leenagupte good point, I didn't explain that sorry, will update.

**Pull request merged**